File: apis/goApi/internal/methods/secant.go
package methods

import (
	"errors"
	"fmt"
	"math"

	"github.com/Jguerra47/NumericalAnalysisProject/apis/goApi/internal/domain"
	"github.com/Jguerra47/NumericalAnalysisProject/apis/goApi/utils"
	"github.com/Knetic/govaluate"
)

func Secant(f string, x0, x1, tol float64, n int) (domain.ResultLine, error) {
	exp, err := govaluate.NewEvaluableExpression(f)
	if err != nil {
		return domain.ResultLine{}, errors.New("Expression mal formed")
	}
	params := make(map[string]interface{})

	ans := domain.ResultLine{}

	params["x"] = x0
	aux, _ := exp.Evaluate(params)
	fx0, _ := aux.(float64)

	if fx0 == 0.0 {
		s := fmt.Sprintf("%v is a root.\n", utils.ConvF(fx0))
		ans.Ans = s
	} else {
		arr := make([]string, 0)
		arr = append(arr, "counter", "xn", "y1", "error")
		ans.Result = append(ans.Result, arr)

		params["x"] = x1
		aux, _ = exp.Evaluate(params)
		fx1, _ := aux.(float64)
		d := (fx1 - fx0)
		errAns := tol + 1
		counter := 1

		arr = make([]string, 0)
		arr = append(arr, fmt.Sprintf("%d", counter), utils.ConvF(x1), utils.ConvF(fx1), utils.ConvF(errAns))
		ans.Result = append(ans.Result, arr)

		for fx1 != 0.0 && errAns > tol && counter < n && d != 0 {
			x2 := x1 - ((fx1 * (x1 - x0)) / d)
			errAns = math.Abs(x2 - x1)
			x0 = x1
			fx0 = fx1

			params["x"] = x2
			aux, _ = exp.Evaluate(params)
			fx1, _ = aux.(float64)

			x1 = x2
			d = (fx1 - fx0)
			counter++

			arr = make([]string, 0)
			arr = append(arr, fmt.Sprintf("%d", counter), utils.ConvF(x1), utils.ConvF(fx1), utils.ConvF(errAns))
			ans.Result = append(ans.Result, arr)
		}

		if fx1 == 0.0 {
			s := fmt.Sprintf("%v is a root.\n", utils.ConvF(x1))
			ans.Ans = s
		} else if errAns < tol {
			s := fmt.Sprintf("%v is an approximation with tolerance %v.\n", utils.ConvF(x1), tol)
			ans.Ans = s
		} else if d == 0.0 {
			s := fmt.Sprintf("denominator is zero, fail.")
			ans.Ans = s
		} else {
			s := fmt.Sprintf("The method fails in iterarion %d.\n", counter)
			ans.Ans = s
		}
	}
	return ans, nil
}

File: NumericSquadProject/numericApp/methods/Roots/secant.py
import sympy as sm 

x = sm.symbols('x')

def secant(f, x0, x1, tol, nIter):
    f = sm.sympify(f)
    fx0 = float(f.subs(x,x0))
    fx1 = float(f.subs(x, x1))
    matrix = []
    if(fx0 == 0):
        return(f"%E is root"%(x0)),[]
    
    else:
        cont = 1
        error = tol+1
        den = fx1-fx0
        # print(f'%5s | %20E | %20E'%(-1,x0,error))
        matrix.append([0,x0,error])
        while(error >= tol and fx1 != 0 and den != 0 and cont < nIter):
            
            x2 = x1-(fx1*(x1-x0))/den

            error = abs(x2-x1)
            
            x0 = x1
            x1 = x2
            # print(f'%5s | %20E | %20E'%(cont,x0,error))
            matrix.append([cont,x0,error])
            #print(matrix[-1])
            fx0 = float(fx1)
            fx1 = float(f.subs(x,x1))
            
            den = fx1-fx0
            cont = cont + 1
        
        if(fx1 == 0):
            ans = (f"%E is a root"%(x1))
                
        elif(error < tol):
            ans = (f"%E is an approximation to a root with a tolerance of %E" %(x1,tol))
                
        elif(den == 0):
            ans = ("There is a possible multiple root")
            
        else:
            ans = ("Failed in " + str(nIter) +  " iterations")
        return ans,matrix



# f = sm.sympify("log(sin(x)^2 + 1)-(1/2)")
# tol = 1e-7
# x0 = 0.5
# x1 = 1
# nIter = 100
# print(secant(f,x0,x1,tol,nIter))